src/planner/planner.c

```diff
@@ -406,6 +406,20 @@ preprocess_query(Node *node, PreprocessQueryContext *context)
 
 					if (ht)
 					{
+						/* SELECT 1 FROM ONLY "foo"."bar" x WHERE "time" OPERATOR(pg_catalog.=) $1
+						 * FOR KEY SHARE OF x */
+						if (query->commandType == CMD_SELECT && query->hasForUpdate &&
+							list_length(query->rtable) == 1 && context->root->glob->boundParams)
+						{
+							RangeTblEntry *rte = linitial_node(RangeTblEntry, query->rtable);
+							if (!rte->inh && rte->rellockmode == RowShareLock &&
+								list_length(query->jointree->fromlist) == 1 &&
+								query->jointree->quals && strcmp(rte->eref->aliasname, "x") == 0)
```

Comment:
Is the name "x" something we can rely on? Doesn't seem as static as you make it out here. For instance, what if somebody uses "y"?

Reply:
Also, this needs a test case.

Reply:
This is tested implicitly by the fk tests.